[Style] Convert position types to use strong style types

weinig opened the pull request:
#### cae98906936e9031add5a63bc95b9473e8ddd1b8
<pre>
[Style] Convert position types to use strong style types
<a href="https://bugs.webkit.org/show_bug.cgi?id=295792">https://bugs.webkit.org/show_bug.cgi?id=295792</a>

Reviewed by Darin Adler.

Adds strong style types for the `object-position`, `background-position`,
`mask-position`, `perspective-origin` and `transform-origin` properties.

Updates existing `Style::Position` to use shared `Style::LengthWrapperBase`
types allowing for the removal of a bunch of conversion code.

* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSToStyleMap.cpp:
* Source/WebCore/css/scripts/process-css-properties.py:
* Source/WebCore/platform/animation/AcceleratedEffectValues.cpp:
* Source/WebCore/rendering/BackgroundPainter.cpp:
* Source/WebCore/rendering/RenderImage.cpp:
* Source/WebCore/rendering/RenderLayer.cpp:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
* Source/WebCore/rendering/RenderReplaced.cpp:
* Source/WebCore/rendering/style/FillLayer.cpp:
* Source/WebCore/rendering/style/FillLayer.h:
* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/RenderStyleSetters.h:
* Source/WebCore/rendering/style/StyleMiscNonInheritedData.h:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.cpp:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/rendering/style/StyleTransformData.cpp:
* Source/WebCore/rendering/style/StyleTransformData.h:
* Source/WebCore/rendering/updating/RenderTreeBuilderSVG.cpp:
* Source/WebCore/style/StyleBuilderConverter.h:
* Source/WebCore/style/StyleBuilderCustom.h:
* Source/WebCore/style/StyleExtractorConverter.h:
* Source/WebCore/style/StyleExtractorCustom.h:
* Source/WebCore/style/StyleExtractorSerializer.h:
* Source/WebCore/style/StyleInterpolationWrappers.h:
* Source/WebCore/style/values/images/StyleGradient.cpp:
* Source/WebCore/style/values/images/StyleObjectPosition.h: Added.
* Source/WebCore/style/values/primitives/StyleLengthWrapper.h:
* Source/WebCore/style/values/primitives/StylePosition.cpp:
* Source/WebCore/style/values/primitives/StylePosition.h:
* Source/WebCore/style/values/shapes/StyleShapeFunction.cpp:
* Source/WebCore/style/values/transforms/StylePerspectiveOrigin.h: Added.
* Source/WebCore/style/values/transforms/StyleTransformOrigin.h: Added.
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/297312@main">https://commits.webkit.org/297312@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7a0c59a316c89c94bc981647b37fe9e61a015696

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/111313 "11 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/30979 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/21444 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/117345 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/61581 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/113275 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/31660 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/39561 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/84602 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/114260 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/25284 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/100216 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/65048 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/24626 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/18356 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/61165 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/94666 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/18423 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/120510 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/38362 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/28505 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/93530 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/38738 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/96490 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/93354 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23787 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/38456 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/16228 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/34364 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/38251 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/43728 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/37916 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/41249 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/39618 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/rendering/style/FillLayer.h:
```diff
@@ -166,8 +174,8 @@ class FillLayer : public RefCounted<FillLayer> {
     static FillSize initialFillSize(FillLayerType) { return { }; }
-    static Length initialFillXPosition(FillLayerType) { return Length(0.0f, LengthType::Percent); }
-    static Length initialFillYPosition(FillLayerType) { return Length(0.0f, LengthType::Percent); }
+    static FillPositionX initialFillXPosition(FillLayerType) { return 0_css_percentage; }
+    static FillPositionY initialFillYPosition(FillLayerType) { return 0_css_percentage; }
```
Can these be `constexpr`? Is there any benefit?

weinig replied:
They unfortunately can’t (there is a non constexpr CalculationValue hidden down in the class)

darinadler commented on Source/WebCore/style/StyleBuilderCustom.h:
```diff
@@ -109,6 +110,9 @@ inline OffsetDistance forwardInheritedValue(const OffsetDistance& value) { auto
 inline OffsetPath forwardInheritedValue(const OffsetPath& value) { auto copy = value; return copy; }
 inline OffsetPosition forwardInheritedValue(const OffsetPosition& value) { auto copy = value; return copy; }
 inline OffsetRotate forwardInheritedValue(const OffsetRotate& value) { auto copy = value; return copy; }
+inline Position forwardInheritedValue(const Position& value) { auto copy = value; return copy; }
```
Reading all these functions, I was wondering how they are different from just:
```suggestion
inline Position forwardInheritedValue(const Position& value) { return value; }
```
I suppose copying into a local variable first must have some effect, but what is it?

weinig replied:
Probably none, and I’ve just been copying the idiom.

weinig replied:
Ok, so this is fun. You (!) add this idiom in https://github.com/WebKit/WebKit/commit/4a44702db4c4cfbbcf8e51b1e505ef28825bfb4c#diff-aca64cace236dc9180bbb97adba67b3b91660953ab9b0920d9bdff5635897870R61, but I can't escape blameless, as I reviewed it!

darinadler replied:
Only 8 years ago

weinig replied:
How could we forget!

darinadler commented on Source/WebCore/style/StyleInterpolationWrappers.h:
```diff
@@ -1930,40 +1904,51 @@ class FillLayerWrapper final : public FillLayerWrapperWithGetter<const T&> {
 
-private:
-    bool equals(const FillLayer* a, const FillLayer* b) const final
+    bool equals(const FillLayer* from, const FillLayer* to) const override
```
```suggestion
    bool equals(const FillLayer* from, const FillLayer* to) const final
```

weinig replied:
I would actually love for us to change our standard here. 

The reason to use `override` here (and in places like this) is that the compiler will warn you if you are not in fact overriding something (say, you have a typo in the signature, or the base class changes), and `final` will not.

For example, take the following example.

```c++
class Base {
public:
    virtual int test(const std::string*) const;
};

class Derived1 : public Base {
public:
    virtual int test(const std::string&) const final;
};

class Derived2 : public Base {
public:
    virtual int test(const std::string&) const override;
};
``` 

(https://godbolt.org/z/Kar4jcaP4 to actually run it)

As you can see, only `Derived2` has an error.

darinadler replied:
> The reason to use `override` here (and in places like this) is that the compiler will warn you if you are not in fact overriding something (say, you have a typo in the signature, or the base class changes), and `final` will not.

Wow, that’s a bad surprise.

I remember when @anttijk and I explored `override` vs. `final` vs. `override final` for cases like this years back, and I thought there was no downside whatsoever to using `final` at that time.

We can consider `override final`, because you can write both, but when the class is already `final` I suppose the only reason to write `final` here too is in case of future refactoring that makes the class non-`final`.

darinadler commented on Source/WebCore/style/StyleInterpolationWrappers.h:
```diff
@@ -1930,40 +1904,51 @@ class FillLayerWrapper final : public FillLayerWrapperWithGetter<const T&> {
     }
 
-    void interpolate(FillLayer* destination, const FillLayer* from, const FillLayer* to, const Context& context) const final
+    bool canInterpolate(const FillLayer* from, const FillLayer* to) const override
```
```suggestion
    bool canInterpolate(const FillLayer* from, const FillLayer* to) const final
```

darinadler commented on Source/WebCore/style/StyleInterpolationWrappers.h:
```diff
@@ -1930,40 +1904,51 @@ class FillLayerWrapper final : public FillLayerWrapperWithGetter<const T&> {
+        if (!from || !to)
             return false;
-        return value(a) == value(b);
+        return Style::equalsForBlending(this->value(from), this->value(to));
```
Why is `this->` needed? Asking just this once, but it affects multiple call sites.
```suggestion
        return Style::equalsForBlending(value(from), value(to));
```

weinig replied:
They are actually not needed here anymore, but where when I was trying to inherit from `FillLayerWrapperWithGetter<T>`.

The error you get in that case is:

```
/Users/weinig/Data/Code/WebKit/Source/WebCore/style/StyleInterpolationWrappers.h:1967:50: error: use of undeclared identifier 'value'
 1967 |         (destination->*this->m_setter)(blendFunc(value(from), this->value(to), context));
```

One of those c++ lookup rules things I don't quite know the details of fully.

darinadler replied:
Yes, I seem to recall that when your base class is a template then you end up needing those.

darinadler commented on Source/WebCore/style/StyleInterpolationWrappers.h:
```diff
@@ -1930,40 +1904,51 @@ class FillLayerWrapper final : public FillLayerWrapperWithGetter<const T&> {
             return true;
-        if (!a || !b)
+        if (!from || !to)
             return false;
```
Seems a shame to have this logic repeated in derived classes. You’d think we could do this at call sites and have the virtual function take references.

weinig replied:
Well, the goal is for there to be only one of these `Wrapper` classes in the end, so I am not too bothered. (Have already removed about half of them).

darinadler commented on Source/WebCore/style/values/images/StyleGradient.cpp:
```diff
@@ -642,7 +642,7 @@ template<typename GradientAdapter, typename StyleGradient> GradientColorStops co
 }
 
-static inline float positionFromValue(const LengthPercentage<>& coordinate, float widthOrHeight)
+static inline float positionFromValue(LengthWrapperBaseDerived auto const& coordinate, float widthOrHeight)
```
I am not familiar with this syntax. What does the auto do here? Some kind of template thing, but I haven’t seen it before. Also, stylistically we could always put `const&` after the type name, but we typically use `const X&` instead. I suppose there’s a reason we don’t do that here that is related to the `auto` syntax.

weinig replied:
This syntax is the equivalent of saying:

```c++
template<LengthWrapperBaseDerived T>
static inline float positionFromValue(const T& coordinate, float widthOrHeight)
```

But is shorter for cases where you don't need to use the type T anywhere.

I don't believe there is a way to `const` in another position with this syntax.

darinadler commented on Source/WebCore/style/values/primitives/StyleLengthWrapper.h:
```diff
@@ -79,6 +82,10 @@ template<typename Numeric, CSS::PrimitiveKeyword... Ks> struct LengthWrapperBase
 
+    // IPC Support
+    explicit LengthWrapperBase(WebCore::Length::IPCData&& data) : m_value { WTFMove(data) } { validate(m_value); }
+    WebCore::Length::IPCData ipcData() const { return m_value.ipcData(); }
```
Given that we use an rvalue reference in the constructor, which seems to imply that copying has a non-trivial cost, I’m a little surprised that this is a copying getter rather than a lifetime-bound `const IPCData&` return value getter.

weinig replied:
I'm not entirely sure what you are suggesting here. My aim was to forward the falls from WebCore::Length, which have the same signatures.

darinadler replied:
OK. Wonder why `WebCore::Length` is designed that way, but not important right now.

darinadler commented on Source/WebCore/style/values/transforms/StyleTransformOrigin.h:
```diff
@@ -0,0 +1,66 @@
+
+    bool operator==(const TransformOrigin&) const = default;
+};
+template<size_t I> const auto& get(const TransformOrigin& value)
```
```suggestion
};

template<size_t I> const auto& get(const TransformOrigin& value)
```

weinig committed:
[Style] Convert position types to use strong style types
https://bugs.webkit.org/show_bug.cgi?id=295792

Reviewed by Darin Adler.

Adds strong style types for the `object-position`, `background-position`,
`mask-position`, `perspective-origin` and `transform-origin` properties.

Updates existing `Style::Position` to use shared `Style::LengthWrapperBase`
types allowing for the removal of a bunch of conversion code.

* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSToStyleMap.cpp:
* Source/WebCore/css/scripts/process-css-properties.py:
* Source/WebCore/platform/animation/AcceleratedEffectValues.cpp:
* Source/WebCore/rendering/BackgroundPainter.cpp:
* Source/WebCore/rendering/RenderImage.cpp:
* Source/WebCore/rendering/RenderLayer.cpp:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
* Source/WebCore/rendering/RenderReplaced.cpp:
* Source/WebCore/rendering/style/FillLayer.cpp:
* Source/WebCore/rendering/style/FillLayer.h:
* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/RenderStyleSetters.h:
* Source/WebCore/rendering/style/StyleMiscNonInheritedData.h:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.cpp:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/rendering/style/StyleTransformData.cpp:
* Source/WebCore/rendering/style/StyleTransformData.h:
* Source/WebCore/rendering/updating/RenderTreeBuilderSVG.cpp:
* Source/WebCore/style/StyleBuilderConverter.h:
* Source/WebCore/style/StyleBuilderCustom.h:
* Source/WebCore/style/StyleExtractorConverter.h:
* Source/WebCore/style/StyleExtractorCustom.h:
* Source/WebCore/style/StyleExtractorSerializer.h:
* Source/WebCore/style/StyleInterpolationWrappers.h:
* Source/WebCore/style/values/images/StyleGradient.cpp:
* Source/WebCore/style/values/images/StyleObjectPosition.h: Added.
* Source/WebCore/style/values/primitives/StyleLengthWrapper.h:
* Source/WebCore/style/values/primitives/StylePosition.cpp:
* Source/WebCore/style/values/primitives/StylePosition.h:
* Source/WebCore/style/values/shapes/StyleShapeFunction.cpp:
* Source/WebCore/style/values/transforms/StylePerspectiveOrigin.h: Added.
* Source/WebCore/style/values/transforms/StyleTransformOrigin.h: Added.
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/297312@main
